File: src/main/java/com/example/chatService/kafka/GroupMessageOutboxProcessor.java
package com.example.chatService.kafka;

import com.example.chatService.dto.GroupMessageDto;
import com.example.chatService.dto.MessagingStatus;
import com.example.chatService.entity.GroupOutbox;
import com.example.chatService.repository.GroupMessageOutboxRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@Slf4j
@RequiredArgsConstructor
public class GroupMessageOutboxProcessor {

    private final GroupMessageOutboxRepository outboxRepository;
    private final KafkaTemplate<String, GroupMessageDto> kafkaTemplate;

    private static final int BATCH_SIZE = 100;
    private static final String TOPIC = "group-message-topic";

    // 인스턴스 고유 식별자
    @Value("${app.instance-id}")
    private String workerId;

    @Transactional
    @Scheduled(fixedDelay = 50)
    public void processOutbox() {

        // 1) 선점
        //  UPDATE SET status = PROCESSING
        int claimed = outboxRepository.claimBatch(workerId, BATCH_SIZE);
        if (claimed == 0) return;

        List<GroupOutbox> list = outboxRepository
                .findByStatusAndLockedByOrderByIdAsc(MessagingStatus.PROCESSING, workerId);

        for (GroupOutbox box : list) {
            try {
                GroupMessageDto message = GroupMessageDto.builder()
                        .roomId(box.getRoomId())
                        .senderId(box.getSenderId())
                        .senderName(box.getSenderName())
                        .sentAt(box.getEventTimestamp())
                        .content(box.getContent())
                        .build();

                kafkaTemplate.send(TOPIC, box.getRoomId(), message);

                box.setStatus(MessagingStatus.SENT);
                box.setLockedBy(null);
                box.setLockedAt(null);

            } catch (Exception e) {
                // 실패시 복구
                log.error("GroupOutbox processing failed for id=" + box.getId(), e);
                box.setStatus(MessagingStatus.NEW);
                box.setLockedBy(null);
                box.setLockedAt(null);
            }
        }
    }
}

File: src/main/java/com/example/chatService/kafka/GroupMessageOutboxProcessor.kt
package com.example.chatService.kafka

import com.example.chatService.dto.GroupMessageDto
import com.example.chatService.dto.MessagingStatus
import com.example.chatService.repository.GroupMessageOutboxRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class GroupMessageOutboxProcessor(
    private val outboxRepository: GroupMessageOutboxRepository,
    private val kafkaTemplate: KafkaTemplate<String, GroupMessageDto>,
) {
    @Value("\${app.instance-id}")
    private lateinit var workerId: String

    @Transactional
    @Scheduled(fixedDelay = 50)
    fun processOutbox() {
        val claimed = outboxRepository.claimBatch(workerId, BATCH_SIZE)
        if (claimed == 0) {
            return
        }

        val list = outboxRepository.findByStatusAndLockedByOrderByIdAsc(MessagingStatus.PROCESSING, workerId)
        for (box in list) {
            try {
                val message = GroupMessageDto(
                    box.roomId ?: "",
                    box.senderId ?: "",
                    box.senderName ?: "",
                    box.content ?: "",
                    box.eventTimestamp,
                )
                kafkaTemplate.send(TOPIC, box.roomId ?: "", message)
                box.status = MessagingStatus.SENT
                box.lockedBy = null
                box.lockedAt = null
            } catch (e: Exception) {
                log.error("GroupOutbox processing failed for id={}", box.id, e)
                box.status = MessagingStatus.NEW
                box.lockedBy = null
                box.lockedAt = null
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(GroupMessageOutboxProcessor::class.java)
        private const val BATCH_SIZE = 100
        private const val TOPIC = "group-message-topic"
    }
}
